src/charm.py

```diff
@@ -534,20 +534,24 @@ def _execute_migrations(self) -> None:
         if not self._are_relations_ready() or not container.can_connect():
             logger.info("Not ready to execute migrations")
             return
-        if self.model.unit.is_leader():
-            env_settings = self._create_discourse_environment_settings()
-            self.model.unit.status = MaintenanceStatus("Executing migrations")
-            try:
-                migration_process: ExecProcess = container.exec(
-                    [f"{DISCOURSE_PATH}/bin/bundle", "exec", "rake", "--trace", "db:migrate"],
-                    environment=env_settings,
-                    working_dir=DISCOURSE_PATH,
-                    user=CONTAINER_APP_USERNAME,
-                )
-                migration_process.wait_output()
-            except ExecError as cmd_err:
-                logger.exception("Executing migrations failed with code %d.", cmd_err.exit_code)
-                raise
+        env_settings = self._create_discourse_environment_settings()
+        self.model.unit.status = MaintenanceStatus("Executing migrations")
+        # The rails migration task is idempotent and concurrent-safe, from
+        # https://stackoverflow.com/questions/17815769/are-rake-dbcreate-and-rake-dbmigrate-idempotent
+        # and https://github.com/rails/rails/pull/22122
+        # Thus it's safe to run this task on all units to
+        # avoid complications with how juju schedules charm upgrades
+        try:
+            migration_process: ExecProcess = container.exec(
+                [f"{DISCOURSE_PATH}/bin/bundle", "exec", "rake", "--trace", "db:migrate"],
+                environment=env_settings,
+                working_dir=DISCOURSE_PATH,
+                user=CONTAINER_APP_USERNAME,
+            )
+            migration_process.wait_output()
+        except ExecError as cmd_err:
+            logger.exception("Executing migrations failed with code %d.", cmd_err.exit_code)
+            raise
```

Comment:
Should there be any changes to tests to confirm this?

Reply:
Do you mean to confirm that schema updates are idempotent and concurrent-safe? I don't think this is really something we can easily test but I'm also not sure we need to test application functionality in this way. Fwiw we do something similar in the WordPress charm but don't test it there from what I can see.